allocator: Re-create slave keys to recover from identity loss

tgraf opened the pull request:
This PR improves ressiliency when slave keys disappear from an allocator prefix by re-creating slave keys automatically and resolving a small race window.

## Tested manually:
1. Deletion of master key:
```
Oct 16 18:23:31 k8s1 cilium-agent[9312]: level=warning msg="Re-created potentially missing master key" error="<nil>" key=cilium/state/identities/v1/id/55133 subsys=allocator
Oct 16 18:23:31 k8s1 cilium-agent[9312]: level=warning msg="Re-created potentially missing slave key" error="create was unsuccessful" key="cilium/state/identities/v1/value/k8s:io.cilium.k8s.policy.cluster=default;k8s:io.cilium.k8s.policy.serviceaccount=cilium-etcd-operator;k8s:io.kubernetes.pod.namespace=kube-system;k8s:name=cilium-etcd-operator;/192.168.34.11" subsys=allocator
```

2. Deletion of slave key:
```
Oct 16 18:25:43 k8s1 cilium-agent[9312]: level=warning msg="Re-created potentially missing slave key" error="<nil>" key="cilium/state/identities/v1/value/k8s:io.cilium.k8s.policy.cluster=default;k8s:io.cilium.k8s.policy.serviceaccount=cilium-etcd-operator;k8s:io.kubernetes.pod.namespace=kube-system;k8s:name=cilium-etcd-operator;/192.168.34.11" subsys=allocator
```

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5908)
<!-- Reviewable:end -->

tgraf committed:
allocator: Lock master key prefix when reusing existing cluster identity

The master key prefix was only allocated when a master key was freshly
allocated. When re-using a master key, there is a small race window in which
the a remote garbage collector can kick-in and delete the master key in between
the Get() and the creation of the slave key. The race would be immediately
fixed via the master key protection mechanism but it will lead to an
unnecessary delete and re-create event which will trigger unnecessary
regenerations. By locking the master key prefix earlier, the GC can be kept out
during the race window. The cost is a slightly slower allocation.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
allocator: Re-create slave keys when master key is missing

So far, the slave key was assumed to be still present when a missing master key
was detected. In case all slave keys are missing, the master key will be
removed again during the next GC cycle of any node. The behavior is correct but
there is unnecessary churn that is never being recovered from until the identity
is no longer used.

By re-creating the the slave key, recovery is possible and it also increases
resiliency when the kvstore is completely wiped as each node will re-create all
slave keys of locally used keys.

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -477,22 +484,14 @@ func (a *Allocator) lockedAllocate(key AllocatorKey) (ID, bool, error) {
 
-	lock, err := a.lockPath(k)
-	if err != nil {
-		releaseKeyAndID()
```
Shouldn't we still do this?

tgraf replied:
It is moved to the beginning of the function.

tgraf committed:
allocator: test: Disable GC in GC unit tests

The GC is run manually to test correctness. Disable the background GC to
guarantee predictability of the test.

Signed-off-by: Thomas Graf <thomas@cilium.io>
